### PR TITLE
Color calibration : add profiling from color checker

### LINKED ISSUE #5618: Better integration of color corecting software into DT
It's mostly taken from discuss topic so that it won't completely disappear: https://discuss.pixls.us/t/colorchecker-auto-color-luminosity-correction-is-this-currently-possible/17423

**Is your feature request related to a problem? Please describe.**

Now to create new style you need to go between darktable and darktable-chart when using something like Colorchecker, Cmp Digital Target or anything similar.
https://pixls.us/articles/profiling-a-camera-with-darktable-chart/
It have many steps and is slow. At the end we will have many unnecessary profiles that would be used only once.

**Describe the solution you'd like**

1. Select photos (or even only one photo)
2. Trigger option "Profile colors for selected photos"
3. Select photo with color tool (like Colorchecker)
    1. Selecting image could open automatically new window or then could be one more click to trigger it
4. Put grid marks on proper colors (there could be drop down list for known color checker tools, like shown here https://youtu.be/mZvrj8__5r0?t=371 that will show grid with proper number of lines/rows, always open with last used as selected)
    1. Would be even better if image recognition could find it by itself
5. Click "Generate and apply"
6. Done (all images have new color profile applied)

At this point this color profile is useless for any other image so I would prefer if it's not added to any list of styles or similar. It could be stored only with selected images.

**Alternatives**

Not sure, maybe it could also be incorporated into tethering if work on it start again #2930
Probably mostly useful for product and beauty shots.
Take tethered shot with Colorchecker and immediately create color profile from it that will be applied to all following it tethered shots.

**Additional context**

My comment from https://discuss.pixls.us/t/colorchecker-auto-color-luminosity-correction-is-this-currently-possible/17423/16?u=matrixik

> There are quite popular siblings in Poland that are doing weeding photography. In they course about colors ([“Władca barw”](https://sklep.dr5000.com/produkt-MOV0016.html) cost around 10€ but only in Polish) they show how they are using Colorchecker.
> They are photographing it many times per one weeding. Inside church in few places (could have many different light lamps), in the evening (when street lamps have different kind of light sources), in the hall.
> Screenshot from 2020-04-10 11-43-54
> They use global profiles (dual illuminant, with custom changes they would do inside Lightroom almost every time, in they words “profile from profile, changing proper colors to nice looking colors”) to make photos from Nikon, Olympus, Fuji and Sony looking the same and then using individual profiles for group of photos from each event (or even only for one photo). Thanks to this most of they photos they simply mass import to Lightroom and then export to jpeg without doing anything. If there is need then they use tool from x-rite to generate profile from Colorchecker and apply it only to few photos that need it, standard adjustments and done. Then they could forget about this profile.
> So they can change this picture
> Screenshot from 2020-04-10 12-43-14
> into this with few clicks
> Screenshot from 2020-04-10 12-43-21
> 
> This looks like really fast workflow to me. They are providing they clients with hundreds ready, good looking photos in few days after weeding. Would be nice to have option like “Generate profile/style for individual picture(s)” in dt.
>

### PR DESCRIPTION
**Pull request by aurelienpierre**
Close #5618.

## Purpose 

Add a GUI and a solver that enables to get, from a color checker chart:
1. the exposure compensation and black offset,
2. the illuminant,
3. the 3×3 matrix (equivalent to an input profile) for color mixing

and apply it straight away to color checker. It also lets monitoring the quality of the fit by showing the CIE Delta E 2000. An adaptative solver is used that enables users to best optimize the color fit for:
1. none (non-weighted least-square means) 
2. neutral colors (weighted LSM for low saturation patches)
3. saturated colors (weighted LMS for high saturation patches)
4. skin and soil colors (weighted LMS for hues around 1 rad in LCh)
5. foliage (weighted LMS for hues around 2.23 rad in LCh)
6. water and sky colors (weighted LMS for hues around -1.93 rad in LCh)

## Supported color checkers
1. Spyder Checkr 48 patches,
2. Xrite Color Checker (Passport, Classic, etc.) 24 patches

## How to use

1. shoot a color checker trying to avoid gloss on the surface, with an even lighting, framing the checker in the center 50% of your image, at f/5.6 minimum and f/8 maximum.
2. open the raw file in darkroom, enable lens correction, disable exposure, use the standard matrix in input color profile and the D65 reference white balance,
3. perform a rough rotation, perspective correction and cropping to ensure the chart is straight and full-frame,
3. in color calibration module, open the collapsible section underneath the tab and choose the appropriate checker in the list,
4. adjust the target on the image by dragging the corner nodes roughly on the reference dots of the chart corners,
5. choose the CAT space in which you want to compute the profiling (a profile is only valid in the CAT space it has been computed in),
6. click the "refresh" button to compute the profile,
7. check the delta E and try other settings (CAT or optimisations) to see which ones minimizes the Delta E,
8. once happy, click on the "validation" button to commit the settings to the module. 

## Screenshots
Screenshot_20201212_211823
Screenshot_20201212_211651

## How to read the chart

A central dot, in the patches, shows the target reference color projected to display color space.

After the first profile is computed (hit the refresh button), the patches display the final Delta E of the corrected image:
1. patches non-crossed mean the final Delta E is < 2.3 (threshold of Just Noticeable Difference)
2. patches crossed with one diagonal mean the final Delta E within 2.3 and 4.6,
3. patches crossed with tow diagonals mean the final Delta is > 4.6.

## Notes

The radius scaling will have an effect on the Delta E, watch out the shadows that often appear around the edges of the patches.

You may reuse a low-delta-E profiles for illuminants that are reasonably close to it, even though the scene has not been shot in the exact same conditions. The matrix will hold and you will only have to change the illuminant depending on the scene. But profiles realized under hardcore lightings, like low-CRI fluorescent bulbs and such are not recommended.

The normalization values are the values you could input, as-is, in the exposure module to get the best fit possible.

A profile is tied to the particular input color profile and WB coefficients in the white balance module. My tests have shown that the lowest delta E is achieved with the defaults of the "modern CAT" workflow.

## Varia

@rabauke if you have some spare time, I would love that you double-check if I haven't done any mistake in the least-squares algo. Also the Delta E 2000 computation needs careful review because the formula is hairy.

### CODE REVIEW

**Comment by MStraeten** [comment]
according to Argyllcms cie files the x-rite Color Checker Passport Lab values differs, so maybe there can be a further Option added:
here the further values for colorchecker.h
```
dt_color_checker_t xrite_PP24 = { .name = "Xrite ColorChecker Passport 24",
                                .author = "Ben Goren <ben@trumpetpower.com>",
                                .date = "Fri Aug  3 15:35:05 MST 2012",
                                .manufacturer = "X-Rite/Gretag Macbeth",
                                .type = COLOR_CHECKER_PASSPORT_XRITE_24,
                                .radius = 0.055f,
                                .ratio = 2.f / 3.f,
                                .patches = 24,
                                .size = { 4, 6 },
                                .middle_grey = 19,
                                .white = 18,
                                .black = 23,
                                .values = { 
                                            { "A1", { 37.972545, 12.84975, 13.905933}, { 0.087, 0.125}},
                                            { "A2", { 65.021429, 17.736538, 18.105244}, { 0.250, 0.125}},
                                            { "A3", { 49.767263, -4.838656, -21.688821}, { 0.417, 0.125}},
                                            { "A4", { 43.171313, -11.673243, 22.59988}, { 0.584, 0.125}},
                                            { "A5", { 55.095608, 8.054, -23.674108}, { 0.751, 0.125}},
                                            { "A6", { 70.345563, -33.87973, 0.591375}, { 0.918, 0.125}},
                                            { "B1", { 62.491292, 36.974257, 57.538878}, { 0.087, 0.375}},
                                            { "B2", { 40.66193, 10.745627, -43.972357}, { 0.250, 0.375}},
                                            { "B3", { 51.418556, 47.83579, 16.229131}, { 0.417, 0.375}},
                                            { "B4", { 31.232269, 21.992198, -20.376399}, { 0.584, 0.375}},
                                            { "B5", { 72.291888, -22.389553, 57.145982}, { 0.751, 0.375}},
                                            { "B6", { 71.992078, 20.037754, 67.243006}, { 0.918, 0.375}},
                                            { "C1", { 28.900464, 11.613838, -47.531575}, { 0.087, 0.625}},
                                            { "C2", { 55.786947, -38.099805, 32.761011}, { 0.250, 0.625}},
                                            { "C3", { 42.323678, 50.646029, 27.808803}, { 0.417, 0.625}},
                                            { "C4", { 81.403423, 3.722966, 80.589317}, { 0.584, 0.625}},
                                            { "C5", { 52.352077, 49.539466, -13.465303}, { 0.751, 0.625}},
                                            { "C6", { 50.742046, -29.398805, -26.412062}, { 0.918, 0.625}},
                                            { "D1", { 96.260066, -0.692605, 4.206323}, { 0.087, 0.875}},
                                            { "D2", { 81.762472, -0.207864, 0.973581}, { 0.250, 0.875}},
                                            { "D3", { 67.22286, -0.574879, 0.364223}, { 0.417, 0.875}},
                                            { "D4", { 51.058515, -0.355289, 0.481164}, { 0.584, 0.875}},
                                            { "D5", { 36.300462, -0.566396, -0.30356}, { 0.751, 0.875}},
                                            { "D6", { 21.283684, 0.258119, 0.046995}, { 0.918, 0.875}} } };
```

**Comment by aurelienpierre** [comment]
@MStraeten awesome, thanks ! I will add them.

**Comment by todd-prior** [comment]
[
[ColorChecker24_After_Nov2014.zip](https://github.com/darktable-org/darktable/files/5685370/ColorChecker24_After_Nov2014.zip)
[ColorChecker24_Before_Nov2014.zip](https://github.com/darktable-org/darktable/files/5685371/ColorChecker24_Before_Nov2014.zip)

](url)

From X-rite  https://www.xrite.com/service-support/new_color_specifications_for_colorchecker_sg_and_classic_charts

**Comment by rabauke** [comment]
@MStraeten As I mentioned earlier, it is _not_ pivotal if a colorchecker passport or some other XRite colorchecker is employed. All colorcheckers use the same 24 reference colors. What matters, however, is the year of production. Reference colors have changed in November 2014 a pointed out by @todd-prior. 

@aurelienpierre Regarding screenshot above: the Color patches that are present in addition to the 24 standard reference colors are not intended to be used for color calibration. This means the left 24 patches in the 1st screenshot above should _not_ be used for color calibration.

**Comment by aurelienpierre** [comment]
> Regarding screenshot above: the Color patches that are present in addition to the 24 standard reference colors are not intended to be used for color calibration. This means the left 24 patches in the 1st screenshot above should _not_ be used for color calibration.

Well, we have them and the reference Lab values, so I see no reason to not use them. They are essentially halftones and more greys, they really don't hurt the delta E.

**Comment by MStraeten** [comment]
ok, here the values before and after 2014:
```
dt_color_checker_t xrite_24_2000 = { .name = "Xrite ColorChecker 24 before 2014",
                                  .author = "X-Rite",
                                  .date = "3/27/2000",
                                  .manufacturer = "X-Rite/Gretag Macbeth",
                                  .type = COLOR_CHECKER_XRITE_24_2000,
                                  .radius = 0.055f,
                                  .ratio = 2.f / 3.f,
                                  .patches = 24,
                                  .size = { 4, 6 },
                                  .middle_grey = 19,   /* 21 for most neutral patch */
                                  .white = 18,
                                  .black = 23,
                                  .values = { 
                                            { "A1", { 37.986,  13.555,  14.059 }, { 0.087, 0.125}},
                                            { "A2", { 65.711,  18.13,   17.81  }, { 0.250, 0.125}},
                                            { "A3", { 49.927, -04.88,  -21.905 }, { 0.417, 0.125}},
                                            { "A4", { 43.139, -13.095,  21.905 }, { 0.584, 0.125}},
                                            { "A5", { 55.112,  08.844, -25.399 }, { 0.751, 0.125}},
                                            { "A6", { 70.719, -33.397,  -0.199 }, { 0.918, 0.125}},
                                            { "B1", { 62.661,  36.067,  57.096 }, { 0.087, 0.375}},
                                            { "B2", { 40.02,   10.41,  -45.964 }, { 0.250, 0.375}},
                                            { "B3", { 51.124,  48.239,  16.248 }, { 0.417, 0.375}},
                                            { "B4", { 30.325,  22.976, -21.587 }, { 0.584, 0.375}},
                                            { "B5", { 72.532, -23.709,  57.255 }, { 0.751, 0.375}},
                                            { "B6", { 71.941,  19.363,  67.857 }, { 0.918, 0.375}},
                                            { "C1", { 28.778,  14.179, -50.297 }, { 0.087, 0.625}},
                                            { "C2", { 55.261, -38.342,  31.37  }, { 0.250, 0.625}},
                                            { "C3", { 42.101,  53.378,  28.19  }, { 0.417, 0.625}},
                                            { "C4", { 81.733,  04.039,  79.819 }, { 0.584, 0.625}},
                                            { "C5", { 51.935,  49.986, -14.574 }, { 0.751, 0.625}},
                                            { "C6", { 51.038, -28.631, -28.638 }, { 0.918, 0.625}},
                                            { "D1", { 96.539,  -0.425,   1.186 }, { 0.087, 0.875}},
                                            { "D2", { 81.257,  -0.638,  -0.335 }, { 0.250, 0.875}},
                                            { "D3", { 66.766,  -0.734,  -0.504 }, { 0.417, 0.875}},
                                            { "D4", { 50.867,  -0.153,  -0.27  }, { 0.584, 0.875}},
                                            { "D5", { 35.656,  -0.421,  -1.231 }, { 0.751, 0.875}},
                                            { "D6", { 20.461,  -0.079,  -0.973 }, { 0.918, 0.875}} } };
                                            
t_color_checker_t xrite_24_2014 = { .name = "Xrite ColorChecker 24 after 2014",
                                  .author = "X-Rite",
                                  .date = "3/28/2015",
                                  .manufacturer = "X-Rite/Gretag Macbeth",
                                  .type = COLOR_CHECKER_XRITE_24_2014,
                                  .radius = 0.055f,
                                  .ratio = 2.f / 3.f,
                                  .patches = 24,
                                  .size = { 4, 6 },
                                  .middle_grey = 19,   /* 21 for most neutral patch */
                                  .white = 18,
                                  .black = 23,
                                  .values = { 
                                            { "A1", { 37.54,   14.37,   14.92 }, { 0.087, 0.125}},
                                            { "A2", { 64.66,   19.27,   17.50 }, { 0.250, 0.125}},
                                            { "A3", { 49.32,  -03.82,  -22.54 }, { 0.417, 0.125}},
                                            { "A4", { 43.46,  -12.74,   22.72 }, { 0.584, 0.125}},
                                            { "A5", { 54.94,   09.61,  -24.79 }, { 0.751, 0.125}},
                                            { "A6", { 70.48,  -32.26,  -00.37 }, { 0.918, 0.125}},
                                            { "B1", { 62.73,   35.83,   56.50 }, { 0.087, 0.375}},
                                            { "B2", { 39.43,   10.75,  -45.17 }, { 0.250, 0.375}},
                                            { "B3", { 50.57,   48.64,   16.67 }, { 0.417, 0.375}},
                                            { "B4", { 30.10,   22.54,  -20.87 }, { 0.584, 0.375}},
                                            { "B5", { 71.77,  -24.13,   58.19 }, { 0.751, 0.375}},
                                            { "B6", { 71.51,   18.24,   67.37 }, { 0.918, 0.375}},
                                            { "C1", { 28.37,   15.42,  -49.80 }, { 0.087, 0.625}},
                                            { "C2", { 54.38,  -39.72,   32.27 }, { 0.250, 0.625}},
                                            { "C3", { 42.43,   51.05,   28.62 }, { 0.417, 0.625}},
                                            { "C4", { 81.80,   02.67,   80.41 }, { 0.584, 0.625}},
                                            { "C5", { 50.63,   51.28,  -14.12 }, { 0.751, 0.625}},
                                            { "C6", { 49.57,  -29.71,  -28.32 }, { 0.918, 0.625}},
                                            { "D1", { 95.19,  -01.03,   02.93 }, { 0.087, 0.875}},
                                            { "D2", { 81.29,  -00.57,   00.44 }, { 0.250, 0.875}},
                                            { "D3", { 66.89,  -00.75,  -00.06 }, { 0.417, 0.875}},
                                            { "D4", { 50.76,  -00.13,   00.14 }, { 0.584, 0.875}},
                                            { "D5", { 35.63,  -00.46,  -00.48 }, { 0.751, 0.875}},
                                            { "D6", { 20.64,   00.07,  -00.46 }, { 0.918, 0.875}} } };
```

**Comment by rabauke** [comment]
@aurelienpierre I must correct myself: The 1st screenshot above shows apparently shows a colorchecker by DataColor which actually has 48 color patches for color calibration. Nevertheless, one needs to distinguish between patches which are designed for color calibration and those for other purposes, e.g., white balancing. Here is an example, where also the white balancing patches are used for calibration: 

Calibration 

More color patches do not automatically give better calibration results. Imagine a color checker with the common 24 fields of 16 saturated color patches and 6 gray patches that is complemented with a large number of desaturated color patches. Depending on the details of the fitting procedure the additional the additional low-saturation fields may decrease the quality of fit for highly saturated colors. As a rule of thumb, it is probably a reasonable idea to employ the same patches for calibration that are used in the manufacturer's software, not more.

**Comment by aurelienpierre** [comment]
> More color patches do not automatically give better calibration results. Imagine a color checker with the common 24 fields of 16 saturated color patches and 6 gray patches that is complemented with a large number of desaturated color patches. Depending on the details of the fitting procedure the additional the additional low-saturation fields may decrease the quality of fit for highly saturated colors. As a rule of thumb, it is probably a reasonable idea to employ the same patches for calibration that are used in the manufacturer's software, not more.

I know, matrix conditionning (conditionnement ?) and that kind of stuff. But my test show that the Datacolor 48 has smaller average delta E than the Xrite ones (which may be explained by the fact that I don't have the time-accurate profile). Anyway, they don't seem to hurt at all.

Thanks @MStraeten, that will be added along with @todd-prior's.

**Comment by MStraeten** [comment]
i simply reordered the patches of the files provided by @todd-prior (A-F --> 1-6, 1-4 --> A-D) to match the current scheme - so according to the clarification from @rabauke no need to differentiate between colorchecker plain & passport - just by production date

**Comment by elstoc** [comment]
Would be nice to get the [Datacolor SpyderCHECKR 24 - SCK200](https://www.datacolor.com/wp-content/uploads/2018/01/SpyderCheckr_Color_Data_V2.pdf) supported as well as it looks like a reasonably priced alternative to the x-rite. Not sure how to get this colour data into the format needed by the module though.

**Comment by todd-prior** [comment]
Aurelien has the data. He had asked me to configure it…I think it will be included…

**Comment by todd-prior** [comment]
@elstoc   see most recent posts..   https://discuss.pixls.us/t/introducing-color-calibration-module-formerly-known-as-channel-mixer-rgb/21227/124

**Comment by elstoc** [comment]
@todd-prior thanks. It must have been too abbreviated (SC24) for my eyes to focus on it :)

**Comment by todd-prior** [comment]
So easy to miss my fault





Get Outlook for Android

**Comment by cryptomilk** [comment]
@aurelienpierre Does the datamodel also support more patches? Could I provide a t_color_checker_t for a ColorCheckerSG and it would work? Or is the patch field limited?

**Comment by johnny-bit** [comment]
@cryptomilk - seems like `dt_color_checker_t.values` accepts any number of values...  just make sure `.patches` and then number of items in `.values` is correct :)

**Comment by aurelienpierre** [comment]
> @aurelienpierre Does the datamodel also support more patches? Could I provide a t_color_checker_t for a ColorCheckerSG and it would work? Or is the patch field limited?

No, You can add as many patches as you want, but then the perspective fitting algo might not be accurate enough if patches become really small.

**Comment by seadra** [comment]
This looks amazing!

Will it be possible for users to add new color chart data (like [this one](https://www.amazon.com/Andoer-Professional-Superior-Digital-Correction/dp/B00ZCHHDOW/ref=cm_cr_arp_d_product_top?ie=UTF8))?

**Comment by aurelienpierre** [comment]
> Will it be possible for users to add new color chart data (like this one)?

No. This chart looks like chinese crap, it's not possible to manufacture accurate color patches at such a low price. Plus the manufacturer doesn't seem to provide the reference CIE Lab or XYZ colors for the patches, so you can't calibrate anything. Only Xrite and Spyder provide consistent color patches.

**Comment by aurelienpierre** [comment]
The latest commit adds support for Spyder 24 and Xrite pre- and post- 2014

**Comment by todd-prior** [comment]
Thanks for the update





Get Outlook for Android

**Comment by seadra** [comment]
> No. This chart looks like chinese crap, it's not possible to manufacture accurate color patches at such a low price. Plus the manufacturer doesn't seem to provide the reference CIE Lab or XYZ colors for the patches, so you can't calibrate anything. Only Xrite and Spyder provide consistent color patches.


They actually provide the reference RGB colors.

I understand what you're getting at, but what I'm asking isn't tied to a specific product (it can be a super luxury $100000 ivory-golden frame French masterpiece color chart if you prefer): from a software point of view, it would be helpful to have user-defined color-charts.

**Comment by johnny-bit** [comment]
@seadra not really. There is no reason to support this particular color checker as it's almost guaranteed to be "off". RGB reference colors are actually bad thing because " Which exactly RGB space are those? " You only want to support we'll known, repeatable checkers for color sensitive stuff, otherwise you might as well eyeball it and be done ;) there are also couple issues with the build of that checker - the patches aren't uniformly square which limits area of checking and we don't know whether color path is made using proper finish (needs to be sorta mat) 

As for "user provided" Part - you can always maintain local patch that has your reference cc. And if there comes color checker maker with new cc that is good in quality and provides proper reference data I'm sure nobody will oppose including it.

**Comment by da-phil** [comment]
@aurelienpierre This looks already pretty nifty,thanks for the work so far ;)
Once you did the calibration procedure you described in the "How to use" section, how could you save away the identified profile? As a preset of the module, I'd assume? 
What do you think of the approach of saving an ICC profile away after calibration and using it as an "input color profile" such as in this 3rd party color calibration plugin (see [video](https://www.youtube.com/watch?v=SHGrALNhWd8))?
I kind of like this approach, but I'm not sure if this is rather abusing the "input color profile" setting?
Or maybe as an input to the "lut 3D" module?

**Comment by seadra** [comment]
@johnny-bit I agree there is no reason to add that particular color checker data ---I actually never asked for it.

I was just hoping that the users will be allowed provide their own data for whatever color chart they use, be it off-market or DIY ---not everyone is willing or has the means to pay for an excessively overpriced piece of painted plastic.

**Comment by MStraeten** [comment]
> What do you think of the approach of saving an ICC profile away after calibration and using it as an "input color profile" such as in this 3rd party color calibration plugin (see [video](https://www.youtube.com/watch?v=SHGrALNhWd8))?
> I kind of like this approach, but I'm not sure if this is rather abusing the "input color profile" setting?
> Or maybe as an input to the "lut 3D" module?
That doesn’t make sense, since the calculation is based on the darktable stuff earlier in the pipe - so it’s not valid for a third party software or later in the pipe

**Comment by parafin** [comment]
> @johnny-bit I agree there is no reason to add that particular color checker data ---I actually never asked for it.
> 
> I was just hoping that the users will be allowed provide their own data for whatever color chart they use, be it off-market or DIY ---not everyone is willing or has the means to pay for an excessively overpriced piece of painted plastic.

How exactly would DIY color checker work? You will have to measure its colors, which is only possible with spectrophotometer, which is also pricey. Then there is a question of glare and color fading. The point is that it's ill-advised and probably not the use-case that should be supported explicitly.

**Comment by johnny-bit** [comment]
> I was just hoping that the users will be allowed provide their own data for whatever color chart they use, be it off-market or DIY ---not everyone is willing or has the means to pay for an excessively overpriced piece of painted plastic.

Lemme explain it better then: The datacolor or pantone (xrite) color checkers are professionally made color patches with well known spectral params with published data used for professional work (mostly). The precision is high enought that semi-gloss finish of Digital SG causes patches that supposed to be "same" as normal color checker 24, are really not the same and have higher deltaE than acceptable if we were to assume those are actually the same.

Given the amount of data we know about color checkers and their spectral params, it's simply not worth to support DIY color checkers or checkers with uncertain characteristics, given that if users were to use them, the deltaE would be too great to even consider. Although as I mentioned previously - this is an open source project, you can create and maintain local patch that will support some cheap piece of painted plastic and will work for you. If there comes another color company to the market with color checker that is consistent like datacolor/xrite ones then there won't be any pushback to include those.

And for colorchecker and similar "pro" equipment that's "fancy and expensive": If you need color checker to keep colors consistent, then you can either afford getting one or (in many locations) rent one. I for example only use color checker for product shots were a cost of returns for "it doesn't look like on photos!" (and negative reviews) outweight one-time cost of getting a color checker.

**Comment by aurelienpierre** [comment]
> Once you did the calibration procedure you described in the "How to use" section, how could you save away the identified profile? As a preset of the module, I'd assume?

You get a "commit" button that applies the computed parameters to the module, and indeed you can save those parameters as a preset.

> What do you think of the approach of saving an ICC profile away after calibration and using it as an "input color profile" such as in this 3rd party color calibration plugin (see video)?

It will not work. The parameters are valid only in CAT16 or Bradford space, in which they are computed, and with the current input profile and white balance used in the pipeline. It's highly contextual. If you want general purpose profiles, use [Argyll CMS to produce an ICC file](https://encrypted.pcode.nl/blog/index.html%3Fp=594.html). 

>  Or maybe as an input to the "lut 3D" module?

We don't compute a LUT here, only a matrix and an illuminant.

> I was just hoping that the users will be allowed provide their own data for whatever color chart they use, be it off-market or DIY ---not everyone is willing or has the means to pay for an excessively overpriced piece of painted plastic.

That would only allow user to harm themselves without knowing it. I can add more reference targets manually if needed, but I won't do so for cheap targets. The price you pay with a Spyder or Xrite product is the quality control to ensure pigments match the specified reference for each batch, with a spectrometer, within a certain tolerance range. Even Wolf Faust, who is working seriously, can't get consistent pigments densities, so each batch of targets he does comes with a custom reference file for the values he measured on that batch, again with a spectrometer. That's metrology, you can have reference values all you want, if the manufacturing process is not reliable enough, they are worth nothing. And quality control is expensive. So either buy a decent target from a serious manufacturer, or use the standard input matrix in darktable, but using a bad target is more harmful than doing nothing.

**Comment by cryptomilk** [comment]
```
// helper to deal with patch color
typedef struct dt_color_checker_patch
{
  const char *name;     // mnemonic name for the patch
  float Lab[3];         // reference color in CIE Lab

  // (x, y) position of the patch center, relatively to the guides (white dots)
  // in ratio of the grid dimension along that axis
  struct {
    float x;
    float y;
  };
} dt_color_checker_patch;
```

In what size is x any y? I this centimeter or millimeter?

```
box width: 12.675 mm
box height: 12.75 mm

upper left corner of first patch
x: 27.5 mm
y: 20.0 mm

gap x: 15.175 mm
gap y: 15.25 mm

A1 center x: 27.5 + 12.675/2 = 26.3375 mm
A1 cetner y: 20.0 +  12.75/2 = 26.3750 mm
```

You have e.g. 0.125 in y, is that centimeter. How precise can it be?

**Comment by aurelienpierre** [comment]
> In what size is x any y? I this centimeter or millimeter?

They are expressed as a ratio. 

We create a reference rectangle in-between the reference markers, which is `width`×`height`. Top-left corner gets assigned (x, y) = (0, 0), bottom-right corner gets assigned (x, y) = (1, 1). So any coordinate inside that rectangle is expressed as an offset from the top-left corner, then all x are divided by `width` and all y are divided by `height`.

For example, for the Spyder Color Checkr, width is 297 mm and height is 197 mm. The center of the first patch is at 21 mm from the top-left corner on both axes, so it gets x = 21 / 297 =  0.0707070707070707 and y = 21 / 197 = 0.1065989847715736. Just keep the width, height and positions in the same unit, whatever it is.

Then you round the values at your convenience.

**Comment by matrixik** [comment]
Can [CMP Digital Target](http://www.cmp-color.fr/eng%20digital%20target.html) (that I linked in https://github.com/darktable-org/darktable/issues/5618#issuecomment-648740285) be considered good color checkers or they should be considered cheap knockoffs? (From what I see they are French company focusing on color calibration).

**Comment by aurelienpierre** [comment]
> Can [CMP Digital Target](http://www.cmp-color.fr/eng%20digital%20target.html) (that I linked in [#5618 (comment)](https://github.com/darktable-org/darktable/issues/5618#issuecomment-648740285)) be considered good color checkers or they should be considered cheap knockoffs? (From what I see they are French company focusing on color calibration).

From their website :  
> The measurement file for calibration is provided in spectral mode and the measurements are made for each production batch. for use in the field, the Digital Target mini offers matte surfaces without reflections for easier shooting.

So, answer is no…

**Comment by MStraeten** [comment]
You can add your target definition in src/common/colorchecker.h and build it for your individual use ...

**Comment by johnny-bit** [comment]
> So, answer is no…

Well the short answer is no, long answer would be "while it cannot be provided in code, you can create local patch for yourself that matches your target batch."

File: src/iop/channelmixerrgb.c [hunk 1]
```diff
@@ -988,6 +1025,601 @@ static void check_if_close_to_daylight(const float x, const float y, float *temp
...
+    A[k * 3 * 9 + 18 + 2] = 0.f;
+    A[k * 3 * 9 + 18 + 3] = 0.f;
+    A[k * 3 * 9 + 18 + 4] = 0.f;
+    A[k * 3 * 9 + 18 + 5] = 0.f;
+    A[k * 3 * 9 + 18 + 6] = w * LMS_test[0];
+    A[k * 3 * 9 + 18 + 7] = w * LMS_test[1];
+    A[k * 3 * 9 + 18 + 8] = w * LMS_test[2];
+  }
+
+  fprintf(stdout, "optimize: %i\n", g->optimization);
```
Comment (TurboGit) [inline src/iop/channelmixerrgb.c:1419]: A debug code to remove.

**Comment by TurboGit** [comment]
How is check dE support to work ? I always have a report that it is **bad**:

image

When I click recompute I have:

image

**Comment by StefanDieke** [comment]
This is a great addition to DT! 
 If possible, please include x-rite colorchecker passport (photo 2). It seems to be one of the most widespread color checkers. 
 Highly appreciated!

**Comment by MStraeten** [comment]
latest commit introduces build error on OSX (current master + this pr) :
```
[ 61%] Building C object lib/darktable/plugins/CMakeFiles/channelmixerrgb.dir/introspection_channelmixerrgb.c.o
In file included from /Users/martinstraeten/src/darktable/build/lib/darktable/plugins/introspection_channelmixerrgb.c:157:
/Users/martinstraeten/src/darktable/src/iop/channelmixerrgb.c:2290:1: fatal error: mutex 'g->lock' is not held on every path through
      here [-Wthread-safety-analysis]
}
^
/Users/martinstraeten/src/darktable/src/iop/channelmixerrgb.c:2240:3: note: mutex acquired here
  dt_pthread_mutex_lock(&g->lock);
  ^
1 error generated.
make[2]: *** [lib/darktable/plugins/CMakeFiles/channelmixerrgb.dir/introspection_channelmixerrgb.c.o] Error 1
make[1]: *** [lib/darktable/plugins/CMakeFiles/channelmixerrgb.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by aurelienpierre** [comment]
> How is check dE support to work ? I always have a report that it is bad:

@Turbogit you need to use it on a color checker, fit the grid on it, then start it. Instead of computing new params, it only computes the delta E, but the rest works the same.

Notice that the dE computed from the checking option and the from the profiling options may differ by up to ± 0.05 due to different order of operations.

File: src/iop/CMakeLists.txt [hunk 2]
```diff
@@ -142,7 +142,7 @@ endif(GMIC_FOUND)
 add_iop(toneequal "toneequal.c" DEFAULT_VISIBLE)
 add_iop(filmicrgb "filmicrgb.c")
 add_iop(negadoctor "negadoctor.c")
-add_iop(channelmixerrgb "channelmixerrgb.c")
+add_iop(channelmixerrgb "channelmixerrgb.c" "../chart/common.c")
```
Comment (TurboGit) [inline src/iop/CMakeLists.txt:145]: This looks wrong. The second parameter is a flag (DEFAULT_VISIBLE) to set the default visibility of the module. I don't think you need to add "../chart/common.c" here at all as it is not an iop. If the file is to be compiler it should be added into src/CMakeFile.list.

Comment (TurboGit) [reply src/iop/CMakeLists.txt:145]: It is ok, I was wrong sorry for the noise.

**Comment by TurboGit** [comment]
This will go in 3.6 as not easy to merge. If you really want to have it in 3.4 series, let me know and I'll see what we can do.

**Comment by elstoc** [comment]
Selfishly (because I've just bought a color checker and I don't use master for my main working edits) I'd like it in 3.4.x but I understand the reasoning for leaving it to 3.6

**Comment by da-phil** [comment]
> Selfishly (because I've just bought a color checker and I don't use master for my main working edits) I'd like it in 3.4.x but I understand the reasoning for leaving it to 3.6

Same here :smile: 

But I guess I'll give master a try once the [master snapshots](https://www.darktable.org/install/#3rdparty) contain those very commits.

**Comment by MStraeten** [comment]
> This will go in 3.6 as not easy to merge. If you really want to have it in 3.4 series, let me know and I'll see what we can do.

I included in in a local build of 3.4.x and it works fine. I didn’t update this with the changes made to include it in current master. I can do a diff to see what was changed ...

**Comment by MStraeten** [comment]
here the diff based on commit 9fc939bbfccbb394d3cc3bdd3e562c5c5d2ce8b9 in darktable-3.4.x:

[colorchecker-3.4.x.diff.zip](https://github.com/darktable-org/darktable/files/5766174/colorchecker-3.4.x.diff.zip)

**Comment by Randalix** [comment]
Thanks guys for the amazing work.
Can we also add X-Rite ColorChecker Passport Video?

**Comment by aurelienpierre** [comment]
Yes, if you can provide the Lab or XYZ reference values for the patches and a test picture.

**Comment by todd-prior** [comment]
I think all the colorcheckers use the same data….different layout…this is only for the 24 but it also mentions the passport….



https://xritephoto.com/ph_product_overview.aspx?ID=820&Action=support&SupportID=5159

**Comment by aurelienpierre** [comment]
Notice we also need the geometry of the chart (more importantly : the gap between patches)

**Comment by todd-prior** [comment]
Sorry I didn’t see you had the video version

**Comment by todd-prior** [comment]
Is your routine doing anything other than optimizing delta E??   Then I think all those extra ramp and skin tone patches on the second half are not really going to be used as it is with the normal targets?? Really I have no idea just asking…

**Comment by Randalix** [comment]
This is the [Colormetric for X-Rite ColorChecker Passport Video](https://drive.google.com/file/d/1clFBRm-3fPvddgo_gAjdVTkeuCX_Bjd0/view?usp=sharing)
The gap between the patches is 3mm. X-Total with outter gap is 9cm, Y-Total with outter gap is 6cm.

**Comment by Randalix** [comment]
I would love to help implementing this, but I'm not quite sure what values I need to put in what format.
Iluminant d50 or 65? Do we have all information needed? If we need some more information or I can help somehow -  let me know.

**Pull request merged**